src/Std/Data/DHashMap/Internal/List/Associative.lean

```diff
@@ -2751,128 +2827,223 @@ theorem alterKey_cons_perm [BEq α] [LawfulBEq α] {k : α} {f : Option (β k) 
+      else
+        haveI hc' : containsKey k' l := by rwa [containsKey_alterKey hl, if_neg h] at hc
+        getValueCast k' l hc' := by
+  have := getValueCast?_alterKey  k k' f l hl
+  rw [getValueCast?_eq_some_getValueCast hc] at this
+  split
+  · next heq =>
+    cases eq_of_beq heq
+    apply Option.some_inj.mp
+    simp_all
+  · next heq =>
+    apply Option.some_inj.mp
+    simp_all only [Bool.false_eq_true, Function.comp_apply, dite_false]
+    rw [getValueCast?_eq_some_getValueCast]
+
+theorem getValueCast_alterKey_self (k : α) (f : Option (β k) → Option (β k))
+    (l : List ((a : α) × β a)) (hl : DistinctKeys l) (hc : containsKey k (alterKey k f l)) :
+    haveI hc' : (f (getValueCast? k l)).isSome := by
+      rwa [containsKey_alterKey hl, beq_self_eq_true] at hc
+    getValueCast k (alterKey k f l) hc = (f (getValueCast? k l)).get hc' := by
+  rw [getValueCast_alterKey _ _ _ _ hl]
+  simp
 
-theorem mem_insertEntry_of_key_ne [BEq α] [LawfulBEq α] {a : α} {b : β a}
-    {l : List ((a : α) × β a)} (p : (a : α) × β a)
-    (hne : p.1 ≠ a) : p ∈ insertEntry a b l ↔ p ∈ l := by
-  simp only [insertEntry, cond_eq_if]
+theorem getValueCast!_alterKey {k k' : α} [Inhabited (β k')] {f : Option (β k) → Option (β k)}
+    (l : List ((a : α) × β a)) (hl : DistinctKeys l) : getValueCast! k' (alterKey k f l) =
+      if heq : k == k' then
+        haveI : Inhabited (β k) := ⟨cast (congrArg β <| eq_of_beq heq).symm default⟩
+        cast (congrArg β (eq_of_beq heq)) <| (f (getValueCast? k l)).get!
+      else
+        getValueCast! k' l := by
+  simp only [hl, getValueCast!_eq_getValueCast?, getValueCast?_alterKey, beq_iff_eq,
+    Function.comp_apply]
   split
-  · exact mem_replaceEntry_of_key_ne p hne
-  · simp only [List.mem_cons, or_iff_right_iff_imp, Sigma.ext_iff]
-    exact fun x => hne x.1 |> False.elim
+  · next heq =>
+    cases eq_of_beq heq
+    simp only [cast_eq]
+  · rfl
 
-theorem mem_eraseKey_of_key_ne [BEq α] [LawfulBEq α] {a : α}
-    {l : List ((a : α) × β a)} (p : (a : α) × β a) (hne : p.1 ≠ a) : p ∈ eraseKey a l ↔ p ∈ l := by
-  induction l
-  · simp only [eraseKey_nil]
-  · next ih =>
-    simp only [eraseKey, List.mem_cons]
-    rw [cond_eq_if]
-    split
-    · next h =>
-      rw [iff_or_self, Sigma.ext_iff]
-      exact fun x => (beq_iff_eq.mp h ▸ hne) x.1 |> False.elim
-    · next h =>
-      simp only [List.mem_cons, ih]
+theorem getValueCastD_alterKey {k k' : α} {v : β k'} {f : Option (β k) → Option (β k)}
+    (l : List ((a : α) × β a)) (hl : DistinctKeys l) : getValueCastD k' (alterKey k f l) v =
+      if heq : k == k' then
+        f (getValueCast? k l) |>.map (cast (congrArg β <| eq_of_beq heq)) |>.getD v
+      else
+        getValueCastD k' l v := by
+  simp only [getValueCastD_eq_getValueCast?, hl, getValueCast?_alterKey, beq_iff_eq,
+    Function.comp_apply]
+  split
+  · next heq =>
+    cases eq_of_beq heq
+    simp only [List.cast_eq_id, Option.map_id]
+  · rfl
```

Comment:
Presumably some of these proofs can be golfed. For example, after adding
```lean
@[simp]
theorem Option.map_cast_apply {γ γ' : Type u} (h : γ = γ') (x : Option γ) :
    Option.map (cast h) x = cast (congrArg Option h) x := by
  cases h; cases x <;> simp
```
this proof could be shortened to
```suggestion
  simp only [getValueCastD_eq_getValueCast?, hl, getValueCast?_alterKey, beq_iff_eq,
    Function.comp_apply]
  simp [apply_dite (Option.getD · v)]
```
But it's probably not a good use of time at the moment to golf these proofs too much, as they're not terribly long to begin with. But `apply_dite` is a useful lemma to know.

Reply:
Thanks! I applied `apply_dite` and `apply_ite` where I found it easy to golf.